<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">

    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <title>cApTUrE</title>
  </head>

  <body>
    <header>
      <div class="container">
        <h1>cApTUrE</h1>
        <h2>cApTUrE is a lightweight NGS pipeline, created for theâ€¨ Utah Genome Project (UGP)</h2>

        <section id="downloads">
          <a href="https://github.com/srynobio/cApTUrE/zipball/master" class="btn">Download as .zip</a>
          <a href="https://github.com/srynobio/cApTUrE/tarball/master" class="btn">Download as .tar.gz</a>
          <a href="https://github.com/srynobio/cApTUrE" class="btn btn-github"><span class="icon"></span>View on GitHub</a>
        </section>
      </div>
    </header>

    <div class="container">
      <section id="main_content">
        <h3>

<strong>cApTUrE - vAriant collecTing UGP pipeline</strong>
<br>
<br>

<strong>VERSION</strong>
This document describes version 1.3.0
<br>
<br>

<strong>SYNOPSIS</strong>
<br>
./cApTUrE --config <string> --interval_list <string> --run<br>
./cApTUrE --config file.cfg --interval_list exon_regions.list --file sorted_bams.list --run<br>
./cApTUrE --clean<br>
<br>

<strong>DESCRIPTION</strong>
cApTUrE is a NGS pipeline written in Perl, created for the
<a href="http://weatherby.genetics.utah.edu/UGP/wiki/index.php/Main_Page" target="_blank">Utah Genome Project</a>
<br>
<br>

Currently it incorporates the following tools:
<ul>
	<li>Perl v5.10</li>
	<li>FastQC</li>
  <li>BWA</li>
  <li>Samblaster</li>
  <li>Sambamba</li>
	<li>SAMtools</li>
	<li>GATK 3.0+</li>
</ul>

cApTUrE requires a config file given as the first commandline argument.
GATKs best practices (with some modifications) are followed throughout this pipeline please refer to their site and the <a href="http://weatherby.genetics.utah.edu/UGP/wiki/index.php/Main_Page" target="_blank">UGP wiki</a> for more information.
<br>
<br>

<strong>INSTALLATION</strong>

<em>Perl Modules:</em>
which cApTUrE requires.
<ul>
	<li>Moo</li>
  <li>Parallel::ForkManager</li>
	<li>Config::Std</li>
	<li>IPC::System::Simple</li>
</ul>

<em>External required software:</em>
<ul>
	<li><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">FastQC</a></li>
	<li><a href="http://bio-bwa.sourceforge.net" target="_blank">BWA</a></li>
  <li><a href="https://github.com/GregoryFaust/samblaster" target="_blank">samblaster</a></li>
	<li><a href="http://lomereiter.github.io/sambamba/" target="_blank">Sambamba</a></li>
	<li><a href="http://samtools.sourceforge.net" target="_blank">SAMtools</a></li>
	<li><a href="http://www.broadinstitute.org/gatk/" target="_blank">GATK</a></li>
	<li><a href="http://www.r-project.org/" target="_blank">R</a></li>
</ul>
<strong>CONTENTS</strong>

cApTUrE's file structure:

cApTUrE/data:
<ul>
	<li><em>exome.analysis.sequence.index</em></li>
	<li><em>capture.cfg</em> - This is the main template configure file use with machine_config.  This will create a new template based on file location per server/system.</li>
  <li><em>Region_File</em> - Region list files for both WG and Exome.
</ul>
cApTUrE/bin:
<ul>
	<li><em>cApTUrE</em> - main script</li>
	<li><em>machine_config</em> - Interactive script which creates a hostname.cfg config file, which contains all path locations for a given machine.
	<li><em>project_config</em> - Can use the above generated config file as a template for each dataset to run.
</ul>
cApTUrE/bin/UGP_tools:
<ul>
	<li><em>RegionMaker</em></li> 
	<li><em>Thousand_genome_recreator.pl</em></li>
	<li><em>UGP_Result_Cleanup.pl</em></li>
</ul>

<br><strong>RUNNING CAPTURE:</strong><br>
After downloading and installing all dependences, a typical setup and run would follow these steps:<br>

<br><em><u>Setting up the config file</u>:</em><br>
machine_config has been created to help complete new configure files as needed.
Often many of the values in the config file can be set on a per-machine basis, creating essentially a new master file (hostname.cfg).  Examples of these would be known indel files, VQSR VCFs, BAM background files, and software paths.
project_config is used to create a new project based on the master file information.  The script will output a new .cfg file with fastq paths, output directories, worker number, memory usage and ugp_id.
<br>

<br><em><u>Running cApTUrE</u>:</em><br>
It is recommend that the unix command <a href="http://www.computerhope.com/unix/screen.htm" target="_blank">screen</a> be used. 
When cApTUrE runs it will create a number of log, list, error and report file.  One of these will be PROGRESS.  This file will keep track of each step of the order process, and is one that will be used and reviewed often throughout the pipeline; typically if you have failed runs.  Furthermore, a .log-txt file will be generated which will keep track of the times of each command and the command lines used.
Error tracking is usually done by reviewing error, log, progress and report files.
<br>

<br><strong>INCOMPATIBILITIES</strong><br>
None know, although not tested on Microsoft or OSX.

<br><strong>BUGS AND LIMITATIONS</strong><br>
Please report any bugs or feature requests to:
shawn.rynearson@gmail.com

<br><strong>AUTHOR</strong><br>
Shawn Rynearson &lt;shawn.rynearson@gmail.com&gt;
  </body>
</html>
